### DISCUSSION TITLE
iceoryx Application Memory Management

### DISCUSSION THREAD

**Discussion by elBoberido** [eclipse-iceoryx/iceoryx#1169] [Ideas]
I think all of us know that our application memory management is not ideal, some may even state it is stupid. Basically, we put everything on the stack and regularly blow up the stack. Currently there is also the problem with the C binding and the `do_not_touch_me` buffers with magic numbers.

A fault confessed is half redressed. Here is my proposal for v2.0:
- make all the storage types for the C binding 8 bytes
- the `*_init` function allocate with the system allocator (malloc) and stores the pointer in the storage
- the `*_deinit` deallocate with the system allocator (free)

For v2.1 or 3.0:
- extend `initRuntime` with an optional `MemoryConfig` struct to be able to select an allocator
    - when nothing is specified, the system allocator is used
    - `MemoryConfig` specifies an allocator
        - iceoryx allocator which can take a block of memory from heap, shm, floppy disk, etc
        - it contains fields for the number of `Listener`, `WaitSet`, `ServiceDiscovery`, etc.
        - additionally one can request additional memory for general usage
        - the iceoryx allocator places MemPools for `Listener`, `WaitSets`, etc. in the memory block
        - a bump allocator can be used for the custom memory -> the user could place a mempool for custom types into this memory
- add an `allocate` method to the runtime (alternatively `loan`)
    - `cxx::unique_ptr<T> PoshRuntime::allocate<T>()`, e.g. `auto listener = getRuntime().allocate<Listener>()`
    - `iox_runtime_allocate_listener()` for the C binding
        - deprecated the `*_init`/`*_deinit` functions and make them use `iox_runtime_allocate_listener()` before they are removed at a later release
- alternatively, we could have factories for the types
    - `cxx::unique_ptr<Listener> Listener::allocate()`/`cxx::unique_ptr<Listener> Listener::loan()`
    - `iox_listener_loan()` for the C binding
- all the memory allocation are tracked, even with the system allocator
    - this is useful during development
    - the data is reported to the introspection
    - can be used to create a tailored `MemoryConfig`

@budrus @elfenpiff @MatthiasKillat @FerdinandSpitzschnueffler @dkroenke @mossmaurice what do you think? Shall I create an issue for this?

**Comment by elfenpiff** [comment]
@elBoberido I fully support the plans for 3.0. I think we have too much to do until 2.1 and this seems like something which comes with a lot surprises.

I see the solution for 2.0 not so critical as you do. First of all I agree that the magic numbers in the types are not an ideal solution and that the memory calculation macro for the waitset/listener is far from a suitable. But we have reliable measures in place that make sure we do not shoot ourself in the foot with that.

1. Unit tests which test every struct on every platform if the size and alignment fits. So we are unable to change the waitset/listener size by accident in C++ without correcting the c binding.
2. The macro calculation function is somehow magic but easy to calculate. And again we have a unit test in place that verify that the function is correct and a comment which describes how one can calculate the parameters.

Furthermore, the stack blow up problem would then only be solved for C and not C++ since in C++ we still store the listener/waitset on the stack.

I fear that your 2.0 solution would replace technical debt with another form of technical debt which may be worse. We are aware of the stack problem and mitigated it for now. If the stack of a C users explodes they can always change this code:
```c
iox_listener_storage_t listenerStorage;
iox_listener_t listener = iox_listener_init(&listenerStorage);
```
into
```c
iox_listener_storage_t *listenerStorage = (iox_listener_storage_t*)malloc(sizeof(iox_listener_storage_t));
iox_listener_t listener = iox_listener_init(listenerStorage);
```

But what happens when by some reason someone is relying on the fact that those constructs only use stack memory or what if something else breaks in the code freeze week with this approach? And do we have the time to implement and test this new approach?

With the `struct` approach we provided the user with a choice, either to use the stack or to use the heap if they want - if we follow your quick fix we would force everyone onto the heap and we have to invest time to implement this.

To summarize: The current approach has flaws which we have to solve but it is completely tested and in my opinion we are unable to destroy something by accident when listener/waitset sizes are changing. We do not have to put further work in it.
With your approach we have to take time to implement this - on the first look only a few hours but with the side effect that we force everyone on the heap and maybe we encounter some weird surprises.

**Comment by MatthiasKillat** [comment]
@elBoberido I have only read it quickly and not fully thought about it. I fully I agree we should change something but I would like us to agree on the constraints,  most importantly such as do we want to make it possible to construct these objects on stack. If this is not a requirement everything else is much easier (handles, configurable allocators, factory functions ...). 

Another important consideration is whether we want to keep the `storage_t` struct concept and have a split between storage and handles (e.g. pointers) and storage or if we can just return handles from the library. Unless we want to have some kind of loaning concept (which the current system would be if we would obtain the `storage_t` structs from some kind of allocator, there may be no need for the user to have them (but they increase flexibility and allow for in place construction). Knowing the size of these storage objects (at compile time!) without extra tooling is the problem though.

We can discuss potential solutions once this is more clear, but I agree with you preference with allocator based approaches.
